src/v/storage/tests/compaction_e2e_test.cc

```diff
@@ -224,6 +301,419 @@ TEST_P(CompactionFixtureParamTest, TestDedupeOnePass) {
+
+    auto log_segment_count_before = log->segment_count();
+    // Sanity check we created the right number of segments.
+    // NOTE: ignore the active segment.
+    auto segment_count_before = log_segment_count_before - 1;
+    ASSERT_EQ(segment_count_before, total_segments);
+
+    // Perform first round of sliding window compaction.
+    do_sliding_window_compact(log);
+
+    // Another sanity check after compaction.
+    auto segment_count_after = log->segment_count() - 1;
+    ASSERT_EQ(total_segments, segment_count_after);
+
+    auto summary_after = dir_summary().get();
+    ASSERT_NO_FATAL_FAILURE(summary_after.check_clean(total_segments));
+
+    // The number of duplicates can't exceed the number of segments - 1: the
+    // latest closed segment should have no duplicates, and at worst, each
+    // preceding segment will have 1 duplicate (the last record).
+    {
+        auto consumed_kvs
+          = check_records(cardinality, total_segments - 1).get();
+        ASSERT_NO_FATAL_FAILURE();
+    }
+
+    // Every segment sans the active segment should have a
+    // clean_compact_timestamp set, since we fully indexed all of them.
+    int num_clean_before = 0;
+    for (const auto& seg : log->segments()) {
+        if (seg->index().has_clean_compact_timestamp()) {
+            ++num_clean_before;
+        }
+    }
+    ASSERT_EQ(num_clean_before, total_segments);
+
+    // Compacting again won't attempt again since the segments are marked as
+    // compacted.
+    auto segments_compacted = log->get_probe().get_segments_compacted();
+    do_sliding_window_compact(log);
+    auto segments_compacted_again = log->get_probe().get_segments_compacted();
+    ASSERT_EQ(segments_compacted, segments_compacted_again);
+
+    // Check that the clean_compact_timestamps got persisted in the index_state.
+    int num_clean_again = 0;
+    for (const auto& seg : log->segments()) {
+        if (seg->index().has_clean_compact_timestamp()) {
+            ++num_clean_again;
+        }
+    }
+    ASSERT_EQ(num_clean_again, num_clean_before);
+
+    // Consume again after restarting and ensure our assertions about
+    // duplicates are still valid.
+    restart(should_wipe::no);
+
+    wait_for_leader(ntp).get();
+    partition = app.partition_manager.local().get(ntp).get();
+    log = partition->log().get();
+
+    ASSERT_EQ(log->segment_count(), log_segment_count_before);
+
+    // Check that the clean_compact_timestamps got persisted in the index_state,
+    // even after a restart.
+    int num_clean_after = 0;
+    for (const auto& seg : log->segments()) {
+        if (seg->index().has_clean_compact_timestamp()) {
+            ++num_clean_after;
+        }
+    }
+    ASSERT_EQ(num_clean_after, num_clean_before);
+
+    // Sleep for tombstone.retention.ms time, so that the next time we attempt
+    // to compact the tombstone record will be eligible for deletion.
+    ss::sleep(tombstone_retention_ms).get();
```

Comment:
can we avoid this ss::sleep and switch to using ss::manual_clock instead?

Reply:
[Something like this?](https://gist.github.com/WillemKauf/ea7c28269e2548ee5f79004a3d61e701)